### PR TITLE
Standardize to `ledger.maxAutoAssociations` property limit

### PR DESCRIPTION
**Pull request by tinker-michaelj**
**Description**:
 - Creates a `ledger.maxAutoAssociations` property to manage auto-associations for both accounts and contracts in one place.

### CODE REVIEW

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/context/properties/GlobalDynamicProperties.java [hunk 1]
```diff
@@ -930,4 +933,8 @@ public long maxNumWithHapiSigsAccess() {
     public Set<Address> contractsWithSpecialHapiSigsAccess() {
         return contractsWithSpecialHapiSigsAccess;
     }
+
+    public int maxAutoAssociations() {
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/context/properties/GlobalDynamicProperties.java:937]: `maxAllowedAutoAssociations` ?

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/crypto/HapiCryptoUpdate.java [hunk 2]
```diff
@@ -256,13 +257,18 @@ protected long feeFor(HapiSpec spec, Transaction txn, int numPayerKeys) throws T
...
+                        .setCurrentApproveForAllNftAllowances(Collections.emptySet())
                         .build();
                 var baseMeta = new BaseTransactionMeta(_txn.getMemoBytes().size(), 0);
                 var opMeta = new CryptoUpdateMeta(
                         _txn.getCryptoUpdateAccount(),
                         _txn.getTransactionID().getTransactionValidStart().getSeconds());
                 var accumulator = new UsageAccumulator();
-                cryptoOpsUsage.cryptoUpdateUsage(suFrom(svo), baseMeta, opMeta, ctx, accumulator);
+                // Once account auto-renew is enabled, we could instead leave the explicit auto-assoc lifetime at 0
+                cryptoOpsUsage.cryptoUpdateUsage(suFrom(svo), baseMeta, opMeta, ctx, accumulator, 8000001L);
```
Comment (Neeharika-Sompalli) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/crypto/HapiCryptoUpdate.java:271]: Shouldn't this be 7776000L ?

File: hedera-node/hapi-fees/src/main/java/com/hedera/node/app/hapi/fees/usage/crypto/CryptoOpsUsage.java [hunk 3]
```diff
@@ -122,12 +122,23 @@ public long cryptoAutoRenewRb(final ExtantCryptoContext ctx) {
...
 
+    /**
+     * Returns the estimated resource usage for a crypto update transaction.
+     *
+     * @param sigUsage the already-computed resource usage for signature verification
+     * @param baseMeta the already-computed resource usage for the base transaction
+     * @param cryptoUpdateMeta metadata summarizing the update transaction
+     * @param ctx the current state of the crypto account
+     * @param accumulator the resource usage accumulator
+     * @param explicitAutoAssocSlotLifetime if non-zero, the lifetime to use for resource usage of new auto-renew slots
```
Comment (david-bakin-sl) [inline hedera-node/hapi-fees/src/main/java/com/hedera/node/app/hapi/fees/usage/crypto/CryptoOpsUsage.java:133]: actually only if positive.  if negative doesn't work this way (due to `Math.max`).  maybe on entry check the precondition ≥0?  (Or just hi pass at 0.)

File: hedera-node/hapi-fees/src/main/java/com/hedera/node/app/hapi/fees/pricing/BaseOperationUsage.java [hunk 4]
```diff
@@ -381,8 +381,9 @@ UsageAccumulator cryptoUpdate(final int newAutoAssocSlots) {
                 .build();
         final var cryptoUpdateMeta = new CryptoUpdateMeta(canonicalTxn.getCryptoUpdateAccount(), now);
         final var into = new UsageAccumulator();
+        // Purposely don't include an explicit auto-assoc slot lifetime here
         CRYPTO_OPS_USAGE.cryptoUpdateUsage(
-                SINGLE_SIG_USAGE, NO_MEMO_AND_NO_EXPLICIT_XFERS, cryptoUpdateMeta, ctx, into);
+                SINGLE_SIG_USAGE, NO_MEMO_AND_NO_EXPLICIT_XFERS, cryptoUpdateMeta, ctx, into, 0);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hapi-fees/src/main/java/com/hedera/node/app/hapi/fees/pricing/BaseOperationUsage.java:386]: SHouldn't the base lifetime be 1 month ?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/fees/calculation/utils/AccessorBasedUsages.java [hunk 5]
```diff
@@ -180,7 +182,11 @@ private void estimateCryptoUpdate(
             SigUsage sigUsage, TxnAccessor accessor, BaseTransactionMeta baseMeta, UsageAccumulator into) {
         final var cryptoUpdateMeta = accessor.getSpanMapAccessor().getCryptoUpdateMeta(accessor);
         final var cryptoContext = opUsageCtxHelper.ctxForCryptoUpdate(accessor.getTxn());
-        cryptoOpsUsage.cryptoUpdateUsage(sigUsage, baseMeta, cryptoUpdateMeta, cryptoContext, into);
+        // We use the SDK default auto-renew period for auto-association slot lifetime if account auto-renew is disabled
+        final var explicitAutoRenewSlotLifetime =
```
Comment (david-bakin-sl) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/fees/calculation/utils/AccessorBasedUsages.java:186]: Why is this here instead of, I dunno, `GlobalDynamicProperties`?

Just seems like a unfortunate place to hide this explicit number.

**Review by Neeharika-Sompalli — DISMISSED** [review]
Have a few questions. Otherwise LGTM !

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/crypto/CryptoUpdateTransitionLogic.java [hunk 6]
```diff
@@ -160,6 +160,9 @@ private ResponseCodeEnum sanityCheck(AccountID target, HederaAccountCustomizer c
             if (newMax < ledger.alreadyUsedAutomaticAssociations(target)) {
                 return EXISTING_AUTOMATIC_ASSOCIATIONS_EXCEED_GIVEN_LIMIT;
             }
+            if (newMax > dynamicProperties.maxAutoAssociations()) {
+                return REQUESTED_NUM_AUTOMATIC_ASSOCIATIONS_EXCEEDS_ASSOCIATION_LIMIT;
+            }
             if (dynamicProperties.areTokenAssociationsLimited() && newMax > dynamicProperties.maxTokensPerAccount()) {
                 return REQUESTED_NUM_AUTOMATIC_ASSOCIATIONS_EXCEEDS_ASSOCIATION_LIMIT;
             }
```
Comment (david-bakin-sl) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/crypto/CryptoUpdateTransitionLogic.java:168]: 2nd time I've seen these six lines. Couldn't it go into `CryptoCreateChecks` like `tooManyAutoAssociations` (in the next file)?

**Review by david-bakin-sl — DISMISSED** [review]
ok by me, with some minor comments

**Review by qnswirlds — DISMISSED** [review]
Overall LGTM!

File: hedera-node/hapi-fees/src/main/java/com/hedera/node/app/hapi/fees/pricing/BaseOperationUsage.java (hunk 4 above)
Comment (tinker-michaelj) [reply hedera-node/hapi-fees/src/main/java/com/hedera/node/app/hapi/fees/pricing/BaseOperationUsage.java:386]: The `0` here will leave the base usage computed the same.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/crypto/HapiCryptoUpdate.java (hunk 2 above)
Comment (tinker-michaelj) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/crypto/HapiCryptoUpdate.java]: Indeed...fixed!

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/context/properties/GlobalDynamicProperties.java (hunk 1 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/context/properties/GlobalDynamicProperties.java]: Renamed.

File: hedera-node/hapi-fees/src/main/java/com/hedera/node/app/hapi/fees/usage/crypto/CryptoOpsUsage.java (hunk 3 above)
Comment (tinker-michaelj) [reply hedera-node/hapi-fees/src/main/java/com/hedera/node/app/hapi/fees/usage/crypto/CryptoOpsUsage.java]: Updated javadoc to be more precise.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/fees/calculation/utils/AccessorBasedUsages.java (hunk 5 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/fees/calculation/utils/AccessorBasedUsages.java]: Moved this to `GlobalDynamicProperties`.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/crypto/CryptoUpdateTransitionLogic.java (hunk 6 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/crypto/CryptoUpdateTransitionLogic.java]: Hmm but this is `CryptoUpdate`...not sure how to do this refactor easily.

In modularized codebase, I am making sure such things don't get duplicated. 🙏

**Review by Nana-EC — APPROVED** [review]

**Review by Neeharika-Sompalli — DISMISSED** [review]

**Review by Neeharika-Sompalli — DISMISSED** [review]

**Review by david-bakin-sl — APPROVED** [review]
LGTM

**Review by david-bakin-sl — APPROVED** [review]
LGTM again!

**Review by david-bakin-sl — DISMISSED** [review]
LGTM thrice!

**Review by Neeharika-Sompalli — APPROVED** [review]

**Review by povolev15 — APPROVED** [review]
